tests: benchmark: multicore: Test PWM with low power modes (#17380)
nrfconnect/sdk-nrf: tests/benchmarks/multicore/idle_with_pwm/README.rst

.. _multicore_idle_with_pwm_test:

PWM in low power states test
############################

.. contents::
   :local:
   :depth: 2

This test benchmarks the idle behavior of an application that runs on multiple cores and uses the PWM driver to light up a LED.

The test scenario repeats forever the following:

* Gradually increase PWM duty cycle from 0% to 50% for one second (the LED lights up), power state active.
* Set PWM duty cycle to 0% for one second (PWM disabled, the LED is OFF), power state low.

Requirements
************

The test supports the following development kit:

.. table-from-rows:: /includes/sample_board_rows.txt
   :header: heading
   :rows: nrf54h20dk_nrf54h20_cpuapp

Overview
********

The test demonstrates the use of PWM with low power modes.

The code stored in the :file:`main.c` file is compiled for both application and radio cores.
The application core uses pwm130 to generate PWM signal on **LED2**.
The radio core uses pwm131 to generate PWM signal on GPIO port 0, pin 7.

There are three test configurations in the :file:`testcase.yaml`.

> * ``benchmarks.multicore.idle_with_pwm.nrf54h20dk_cpuapp_cpurad.s2ram'``
>    This configuration uses KConfigs that enable entering low power modes.

Reviewer: ```suggestion
* ``benchmarks.multicore.idle_with_pwm.nrf54h20dk_cpuapp_cpurad.s2ram``
   This configuration uses Kconfigs that enable entering low power modes.
```

   Logging is disabled.

>    Core sleeps for time sufficient to enter the ``suspend-to-ram`` power state.

Reviewer: ```suggestion
   Core sleeps for time sufficient to enter the ``suspend-to-ram`` power state.

```

* ``benchmarks.multicore.idle_with_pwm.nrf54h20dk_cpuapp_cpurad.sleep``

>    This configuration uses KConfigs that enable entering low power modes.

Reviewer: ```suggestion
   This configuration uses Kconfigs that enable entering low power modes.
```

   Logging is disabled.

>    Core sleeps for time sufficient to enter the ``suspend-to-idle``power state.

Reviewer: ```suggestion
   Core sleeps for time sufficient to enter the ``suspend-to-idle``power state.

```

> * ``benchmarks.multicore.idle_with_pwm.nrf54h20dk_cpuapp_cpurad.s2ram'``
>    This configuration uses KConfigs that enable entering low power modes.
>    Logging is disabled.
>    Core sleeps for time sufficient to enter the ``suspend-to-ram`` power state.
> * ``benchmarks.multicore.idle_with_pwm.nrf54h20dk_cpuapp_cpurad.sleep``
>    This configuration uses KConfigs that enable entering low power modes.
>    Logging is disabled.
>    Core sleeps for time sufficient to enter the ``suspend-to-idle``power state.
> * ``benchmarks.multicore.idle_with_pwm.nrf54h20dk_cpuapp_cpurad.no_sleep``
>    You can use this configuration for debug purposes.
>    Logging is enabled while power mode is disabled.

Reviewer: ```suggestion
* ``benchmarks.multicore.idle_with_pwm.nrf54h20dk_cpuapp_cpurad.s2ram``

  This configuration uses Kconfigs that enable entering low power modes.
  Logging is disabled.
  Core sleeps for time sufficient to enter the ``suspend-to-ram`` power state.
  
* ``benchmarks.multicore.idle_with_pwm.nrf54h20dk_cpuapp_cpurad.sleep``

  This configuration uses Kconfigs that enable entering low power modes.
  Logging is disabled.
  Core sleeps for time sufficient to enter the ``suspend-to-idle``power state.
  
* ``benchmarks.multicore.idle_with_pwm.nrf54h20dk_cpuapp_cpurad.no_sleep``

  You can use this configuration for debug purposes.
  Logging is enabled while power mode is disabled.
```

Author: line 42 and 48 contain white character. Is this intentional?

Building and running
********************

.. |test path| replace:: :file:`tests/benchmarks/multicore/idle_with_pwm`

.. include:: /includes/build_and_run_test.txt

To build the test, use configuration setups from the :file:`testcase.yaml` file using the ``-T`` option.
See the example:

> .. code-block:: console
> 
>   west build -p -b nrf54h20dk/nrf54h20/cpuapp -T benchmarks.multicore.idle_with_pwm.nrf54h20dk_cpuapp_cpurad.s2ram .

Reviewer: ```suggestion
.. code-block:: console

   west build -p -b nrf54h20dk/nrf54h20/cpuapp -T benchmarks.multicore.idle_with_pwm.nrf54h20dk_cpuapp_cpurad.s2ram .
```

.. _multicore_idle_with_pwm_test:

PWM in low power states test
############################

.. contents::
   :local:
   :depth: 2

> The test benchmarks the idle behavior of an application that runs on multiple cores and uses PWM driver to light up LED.

Reviewer: ```suggestion
This test benchmarks the idle behavior of an application that runs on multiple cores and uses the PWM driver to light up a LED.
```

> Test scenario
> Repeat forever:
> - gradually light up LED (this lasts for ~1 second),
> - disable PWM (set period to 0); This will turn off LED,
> - sleep for 1 second.

Reviewer: ```suggestion
The test scenario repeats forever the following:

* Gradually light up the LED (this lasts for ~ one second).
* Disable the PWM (set period to 0) and turns off the LED.
* Sleep for one second.
```

Author: I'm not sure about
_Disable the PWM (set period to 0) and turns off the LED._
It's NOT two actions - set period to 0 then switch off LED.
Setting PWM period to 0 is a test step. As a result, LED switches OFF.
My intention was to explain that there are two observable "states":
a) LED lights up from OFF to "bright" which lasts for 1 sec;
b) LED is OFF for 1 sec.
In a) cpu is in power state active and it's using PWM driver.
In b) cpu is in low power state and PWM driver is disabled.

Author: What about
* Disable the PWM (set period to 0, which turns off the LED).
???

Author: I still don't like it.
You can light up / turn off LED with GPIO but this is abrupt change in brightness. Either 0%, either 100%.
Test focuses on enabling / disabling PWM driver. PWM allows to set any brightness between 0-100%.
LED is there only to indicate that PWM is indeed working.

Author: * Gradually increase PWM duty cycle from 0% to 50% in one second (the LED lights up), power state active.
  * Set PWM duty cycle to 0% for one second (PWM disabled, the LED is OFF), power state low.

Reviewer: Well, then how about switching the order like this:
* PWM enabled, the LED lights up for one second (power state active).
* PWM disabled, the LED is turned off (power state low).

Or suggest something simple to outline what happens.

Reviewer: We must have written at the same time. Okay with your suggestion, but *for* one second on the first bullet.

Author: In active state, every 20 ms, PWM duty cycle is increased by 1%. It is done 50 times, so it lasts for 50*20ms = 1 sec.
LED brightness increases from 0 to 50% (uhm, it's not correct; there is non-linear dependency between PWM duty cycle and brightness).
To enter sleep, PWM peripheral has to be disabled. There is no explicit API function. Driver disables PWM peripheral when duty cycle is set to 0% or 100%.

Requirements
************

> The test supports the following development kits:

Reviewer: ```suggestion
The test supports the following development kit:
```

.. table-from-rows:: /includes/sample_board_rows.txt
   :header: heading
   :rows: nrf54h20dk_nrf54h20_cpuapp

Overview
********

> The test demonstrates use of PWM with low power modes.

Reviewer: ```suggestion
The test demonstrates the use of PWM with low power modes.
```

> Code stored in :file:`main.c` is compiled for Application and Radio cores.
> Application core uses pwm130 to generate PWM signal on LED2.
> Radio core uses pwm131 to generate PWM signal on GPIO Port 0 pin 7.

Reviewer: ```suggestion
The code stored in the :file:`main.c` file is compiled for both application and radio cores.
The application core uses pwm130 to generate PWM signal on **LED2**.
The radio core uses pwm131 to generate PWM signal on GPIO port 0, pin 7.
```

There are three test configurations in :file:`testcase.yaml`.

#. 'benchmarks.multicore.idle_with_pwm.nrf54h20dk_cpuapp_cpurad.s2ram'
   This configuration uses KConfigs that enable entering low power modes. However, logging is disabled here.
   Core sleeps for time sufficient to enter "suspend-to-ram" power state.
#. 'benchmarks.multicore.idle_with_pwm.nrf54h20dk_cpuapp_cpurad.sleep'
   This configuration uses KConfigs that enable entering low power modes. However, logging is disabled here.
   Core sleeps for time sufficient to enter "suspend-to-idle" power state.
#. 'benchmarks.multicore.idle_with_pwm.nrf54h20dk_cpuapp_cpurad.no_sleep'
   This configuration can be used for debug purposes. It has enabled logging while Power Moding is disabled.

Building and running
********************

.. |test path| replace:: :file:`tests/benchmarks/multicore/idle_with_pwm`

.. include:: /includes/build_and_run_test.txt

To build the test, use configuration setups from :file:`testcase.yaml` using the ``-T`` option.
See the following examples:

> nRF54H20 DK

Author: OK. I will add headers for two boards when test is aligned to run on nRF54L15 DK.

>   You can build the test for application and radio cores as follows:

Author: Is this line needed then?
I put it here because on 54H it runs on Application and Radio core.
On 54L there is no Radio core so it most likely be Application and FLPR core. (I don't know yet; it's not ready)

Reviewer: You might just leave the cores out. So -> To build the test, use the following command:

  .. code-block:: console

>      west build -p -b nrf54h20dk/nrf54h20/cpuapp -T benchmarks.multicore.idle_with_pwm.nrf54h20dk_cpuapp_cpurad.s2ram .

Author: Right now, it's more important for me that user builds with -T configuration_name option rather what cores are flashed. I will push the changes in a moment...

Testing
=======

After programming the test to your development kit, complete the following steps to test it:

> #. Connect the PPK2 Power Profiler Kit or other current measurement device.
> #. Reset the kit.
> #. Observe the LED2 brightness (or use oscilloscope to collect PWM signal on Port 0 pin 7).
> #. When LED2 is OFF power profiler shall indicate low current consumption resulting from both cores being in low power mode.

Reviewer: ```suggestion
1. Connect the PPK2 Power Profiler Kit or other current measurement device.
#. Reset the kit.
#. Observe the **LED2** brightness (or use oscilloscope to collect the PWM signal on port 0, pin 7).
#. When the **LED2** is off, the power profiler indicates low current consumption resulting from both cores being in low power mode.
```

Author: I see that first line starts with ``1.``, while all other starts with ``#.``
Is this intentional?
**1.** Connect the PPK2 Power Profiler Kit...

Reviewer: Yes, this is intentional. We start the first numbered item with 1. and the rest come just with hashes.
